### PR TITLE
Partial order removal  #714

### LINKED ISSUE #714: Remove part of an order
Currently we can only remove an entire order via the rpc layer. I think it would probably be useful to be able to remove only part of an order by specifying a quantity on the `RemoveOrderRequest`.

### PR DESCRIPTION
**Pull request by rsercano**
Hello,

this hopefully resolves #714, I've tested without **swap** and it's working properly. (without **hold**) But couldn't find a chance to test with **swaps**, aren't they being executed immediately? 

I would love to learn an easy way to hang a swap and test this implementation, without swap as I said it's okay. 

Please let me know for any issues about this PR.

### CODE REVIEW

**Comment by kilrau** [comment]
> I would love to learn an easy way to hang a swap and test this implementation, without swap as I said it's okay.

Good point. Will that be included in https://github.com/ExchangeUnion/xud-simulation ? @moshababo

File: lib/cli/commands/cancelorder.ts [hunk 1]
```diff
@@ -2,18 +2,25 @@ import { callback, loadXudClient } from '../command';
...
 
 export const describe = 'remove an order';
 
 export const builder = {
   order_id: {
     type: 'string',
   },
+  quantity: {
+    type: 'number',
+    describe: 'quantity to remove, this argument is optional',
```
[the remark this answers is not in the document]
Comment (rsercano) [reply lib/cli/commands/cancelorder.ts]: Sure

File: lib/orderbook/OrderBook.ts [hunk 2]
```diff
@@ -447,13 +447,75 @@ class OrderBook extends EventEmitter {
...
-  public removeOwnOrderByLocalId = (localId: string) => {
+  public removeOwnOrderByLocalId = (localId: string, quantityToRemove?: number) => {
     const orderIdentifier = this.localIdMap.get(localId);
     if (!orderIdentifier) {
       throw errors.LOCAL_ID_DOES_NOT_EXIST(localId);
     }
 
     const order = this.getOwnOrder(orderIdentifier.id, orderIdentifier.pairId);
+
+    this.logger.debug(`requested removal of order, quantityToRemove: ${quantityToRemove}, order quantity: ${order.quantity}`);
```
[the remark this answers is not in the document]
Comment (rsercano) [reply lib/orderbook/OrderBook.ts]: The approach you've mentioned is okay to me, I personally log every entry and end point of methods as a habit :) So, as you've mentioned there's no actual point to print undefined often.

File: lib/orderbook/OrderBook.ts [hunk 3]
```diff
@@ -447,13 +447,75 @@ class OrderBook extends EventEmitter {
...
+    if (quantityToRemove) {
+      this.removeOwnOrderByLocalIdWithQuantity(orderIdentifier, order, quantityToRemove, localId);
+    } else {
+      this.removeOwnOrderByLocalIdWithoutQuantity(orderIdentifier, order, localId);
+    }
+
+    return order.hold;
+  }
+
+  private removeOwnOrderByLocalIdWithQuantity = (orderIdentifier: OrderIdentifier, order: OwnOrder, quantityToRemove: number, localId: string) => {
```
[the remark this answers is not in the document]
Comment (rsercano) [reply lib/orderbook/OrderBook.ts]: In my opinion, it's better to keep those different logic separately for the sake of readability. They're not completely different but they don't have much in common. But you're right about duplication, may I suggest to put all duplicate code into a single function ?

File: lib/orderbook/OrderBook.ts [hunk 4]
```diff
@@ -447,13 +447,75 @@ class OrderBook extends EventEmitter {
...
+    if (quantityToRemove > order.quantity) {
+      // quantity to be removed can't be higher than order's quantity.
+      throw errors.QUANTITY_DOES_NOT_MATCH(quantityToRemove, order.quantity);
+    } else if (quantityToRemove === order.quantity) {
+      return this.removeOwnOrderByLocalIdWithoutQuantity(orderIdentifier, order, localId);
+    }
+
+    const quantityCanBeRemoved = order.quantity - (order.hold ? order.hold : 0);
```
[the remark this answers is not in the document]
Comment (rsercano) [reply lib/orderbook/OrderBook.ts]: done !

File: lib/orderbook/OrderBook.ts [hunk 5]
```diff
@@ -447,13 +447,75 @@ class OrderBook extends EventEmitter {
...
+    if (quantityToRemove <= quantityCanBeRemoved) {
+      this.removeOwnOrder(orderIdentifier.id, orderIdentifier.pairId, quantityToRemove);
+    }else {
```
[the remark this answers is not in the document]
Comment (rsercano) [reply lib/orderbook/OrderBook.ts]: Sure, I guess there's no tslint rule currently for this, right ?

File: lib/orderbook/OrderBook.ts [hunk 6]
```diff
@@ -447,13 +447,75 @@ class OrderBook extends EventEmitter {
...
+        }
+      };
+
+      const failedHandler = (deal: SwapDeal) => {
+        if (deal.orderId === orderIdentifier.id) {
+          // remove the portion that failed now that it's not on hold
+          if (deal.quantity! > remainingToRemove) {
+            this.removeOwnOrder(orderIdentifier.id, orderIdentifier.pairId, remainingToRemove);
+            cleanup(remainingToRemove);
+          }else {
```
[the remark this answers is not in the document]
Comment (rsercano) [reply lib/orderbook/OrderBook.ts]: sure

File: lib/orderbook/OrderBook.ts [hunk 7]
```diff
@@ -447,13 +447,75 @@ class OrderBook extends EventEmitter {
...
+    } else if (quantityToRemove === order.quantity) {
+      return this.removeOwnOrderByLocalIdWithoutQuantity(orderIdentifier, order, localId);
+    }
+
+    const quantityCanBeRemoved = order.quantity - (order.hold ? order.hold : 0);
+    if (quantityToRemove <= quantityCanBeRemoved) {
+      this.removeOwnOrder(orderIdentifier.id, orderIdentifier.pairId, quantityToRemove);
+    }else {
+      this.removeOwnOrder(orderIdentifier.id, orderIdentifier.pairId, quantityCanBeRemoved);
+      let remainingToRemove = quantityToRemove - quantityCanBeRemoved;
```
[the remark this answers is not in the document]
Comment (rsercano) [reply lib/orderbook/OrderBook.ts]: Of course, done.

**Comment by moshababo** [comment]
>> I would love to learn an easy way to hang a swap and test this implementation, without swap as I said it's okay.

>Good point. Will that be included in https://github.com/ExchangeUnion/xud-simulation ? @moshababo 

I don't see an easy way to control this via RPC. We can do a hack if needed..

**Comment by moshababo** [comment]
>> I would love to learn an easy way to hang a swap and test this implementation, without swap as I said it's okay.

>Good point. Will that be included in https://github.com/ExchangeUnion/xud-simulation ? @moshababo 

I don't see an easy way to control this via RPC. We can do a hack if necessary.

File: lib/orderbook/OrderBook.ts [hunk 8]
```diff
@@ -447,13 +447,75 @@ class OrderBook extends EventEmitter {
...
+      this.removeOwnOrderByLocalIdWithoutQuantity(orderIdentifier, order, localId);
+    }
+
+    return order.hold;
+  }
+
+  private removeOwnOrderByLocalIdWithQuantity = (orderIdentifier: OrderIdentifier, order: OwnOrder, quantityToRemove: number, localId: string) => {
+    if (quantityToRemove > order.quantity) {
+      // quantity to be removed can't be higher than order's quantity.
+      throw errors.QUANTITY_DOES_NOT_MATCH(quantityToRemove, order.quantity);
```
[the remark this answers is not in the document]
Comment (moshababo) [reply lib/orderbook/OrderBook.ts]: IMHO an error is preferred - it might help to discover a bug on the client.

**Comment by kilrau** [comment]
Passes your review now? @sangaman

File: lib/orderbook/OrderBook.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (sangaman) [reply lib/orderbook/OrderBook.ts]: Yes, but after some digging I found the rule we want and I'll open a separate PR for it. Thanks.

File: lib/orderbook/OrderBook.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (sangaman) [reply lib/orderbook/OrderBook.ts]: Yeah if you have an approach in mind where there are separate methods without code duplication that'd probably be fine with me. I'll take another look after you commit it.

**Comment by sangaman** [comment]
Just one last item, I marked the rest as resolved. @rsercano I'll review again as soon as you commit, thanks a lot.

File: lib/orderbook/OrderBook.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (rsercano) [reply lib/orderbook/OrderBook.ts]: Sure thing, I'll check it today afternoon

[the remark this answers is not in the document]
Comment (rsercano) [reply lib/orderbook/OrderBook.ts]: I couldn't find enough time to check this in detail, very sorry about it, just quickly pushed a commit which prevents most of the code duplication (didn't like the code I wrote :-1: )

**Comment by rsercano** [comment]
Very sorry, couldn't check the last point @sangaman has mentioned in detail having really busy days, just pushed a commit which prevents most of the duplication but it's really not good and `proceedRemovingOwnOrder` method need to be reworked (working but not really well designed in terms of coding)

File: lib/orderbook/OrderBook.ts [hunk 9]
```diff
@@ -479,51 +479,94 @@ class OrderBook extends EventEmitter {
...
+        this.logger.debug(`can't remove local order ${localId} yet because it has a hold of ${order.hold}`);
+        const availableQuantity = order.quantity - order.hold;
+        if (availableQuantity) {
+          this.removeOwnOrder(orderIdentifier.id, orderIdentifier.pairId, availableQuantity);
         }
-      };
 
-      this.swaps!.on('swap.failed', failedHandler);
-      this.swaps!.on('swap.paid', paidHandler);
+        failedHandler = (deal: SwapDeal) => {
```
Comment (sangaman) [inline lib/orderbook/OrderBook.ts]: Per my comment above, I'd suggest dropping this additional `failedHandler` definition and only use the one defined above that checks against `remainingQuantityToRemove`, and move it back near the top of the method as a `const`.

**Comment by sangaman** [comment]
Per our discussion on the call, I'm going to move these changes into a separate branch/PR and implement my latest feedback so we can get it in for the next release.

**Comment by sangaman** [comment]
Closing this because #803 was merged which had the same changes.

**Pull request closed without merging**